src/System.Memory/ref/System.Memory.cs

```diff
@@ -508,4 +509,12 @@ public static partial class MemoryMarshal
         public static System.Span<T> CreateSpan<T>(ref T reference, int length) { throw null; }
 #endif
     }
+
+    public static partial class SequenceMarshal
```

Comment:
Why do we have separate SequenceMarshal class? Can we use the existing MemoryMarshal?

Reply:
That is exactly the question that I was going to ask you :) see https://github.com/dotnet/corefx/pull/27229#issuecomment-367049554

Reply:
I recommend adding it on the coreclr side first (which will get mirrored here, in corefx). Only then we will get the tests passing.

See the example of API addition here (in the example @benaadams added TryGetOwnedMemory APIs to MemoryMarshal):
1) Add new API to coreclr (https://github.com/dotnet/coreclr/pull/16455).
2) The mirror PR will automatically get created to add the change on the corefx side (https://github.com/dotnet/corefx/pull/27325).
3) Wait for a versions-update PR to go through, which points corefx to the coreclr with your change (https://github.com/dotnet/corefx/pull/27241).
4) Expose the API in corefx and add tests (https://github.com/dotnet/corefx/pull/27288).

> couldn't fix the build because sometimes it comes from `System.Private.CoreLib`

On netfx, your change here will likely pass since it doesn't depend on coreclr. The other legs would fail though.

Reply:
But ReadOnlySequence is not in coreclr at all.

Reply:
Ah yes, good point!

MemoryMarshal is a partial class ([common](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Runtime/InteropServices/MemoryMarshal.cs), [portable](https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.Portable.cs), [fast](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Runtime/InteropServices/MemoryMarshal.Fast.cs)) and its surface area from corelib has to be a superset of what's coming from System.Memory (it contains some fast span specific APIs):
https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs#L502

Therefore, atm, if you add an API to it, it must be implemented in coreclr. 

It is a bit tricky, but what if we add these to MemoryMarshal.Sequence.cs (which defines the partial MemoryMarshal class) and include that file here (if IsPartialFacadeAssembly == true):
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System.Memory.csproj#L153

Any concerns with that approach?

Reply:
Yeah there is a problem. I talked to @weshaggard about this and to make MemoryMarshal work we would need to mirror everything related to ReadOnlySequence to coreclr because system.memory for netcoreapp is entirely built from there.

Reply:
System.Memory for netcoreapp is not *entirely* built from coreclr. We still have types that come from System.Memory. See every unconditional ItemGroup: https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System.Memory.csproj#L24-L108

Reply:
@KrzysztofCwalina and I chatted this morning and given the fact we don't want  to move ReadOnlySequence to coreclr we are going to need a new type for these so they can stay just in System.Memory. @KrzysztofCwalina wanted to come up with a better name then SequenceMarshal but I suppose for the purpose of this PR we can stick with that and file a new issue on the naming of the type.

Reply:
> we don't want to move ReadOnlySequence to coreclr we are going to need a new type for these so they can stay just in System.Memory

For it to stay in System.Memory, do we have to create a new type? Can you explain why using MemoryMarshal would force ReadOnlySequence to move to coreclr (which I understand that we don't want)?

Reply:
>System.Memory for netcoreapp is not entirely built from coreclr. We still have types that come from System.Memory. See every unconditional ItemGroup: 

Not all of System.Memory but all of MemoryMarshal is implemented in coreclr for netcoreapp.

Reply:
> Not all of System.Memory but all of MemoryMarshal is implemented in coreclr for netcoreapp.

My suggestion [here](https://github.com/dotnet/corefx/pull/27229#discussion_r170143939) suggests an approach to change that.

Reply:
That approach will not work. It would only work if you plan to expose these APIs only in the OOB implementation which we do not want to do. MemoryMarshal implementation cannot be split between two assemblies.

Reply:
> MemoryMarshal implementation cannot be split between two assemblies.

Yes, you are right. Ignore my suggestion. We will have to use a separate class for this. Thanks for the clarification.